File: Wells/Source/VievModels/FilterViewModel.vb
Imports System.Reflection
Imports Wells.ViewBase
Imports Wells.Base.Common
Public Class FilterViewModel
    Inherits BaseViewModel

    Private _Properties As Dictionary(Of String, PropertyInfo)
    Private _SelectedPropertyIndex As Integer

    Property ShowStringPanel As Boolean
    Property ShowNumericPanel As Boolean
    Property ShowBooleanPanel As Boolean
    Property ShowDatePanel As Boolean
    Property ShowEnumPanel As Boolean
    Property NumericValue As Double
    Property StringValue As String
    Property SelectedEnumValue As Integer
    Property StartDate As Date
    Property EndDate As Date
    Property BooleanValue As Boolean
    Property FilterType As Type
    Property SelectedMathFunction As Integer
    Property PropertyName As String
    Property PropertyDisplayName As String
    Property IsIntegerNumericType As Boolean
    Property IsCreatingFilter As Boolean
    Property DialogTitle As String

    Sub New(properties As Dictionary(Of String, PropertyInfo))
        MyBase.New(Nothing)
        _Properties = properties
        EndDate = Date.Today
        BooleanValue = True
        SelectedEnumValue = 0
        IsIntegerNumericType = False
        IsCreatingFilter = True
        If properties.Any Then
            SelectedPropertyIndex = 0
        End If
        DialogTitle = "Crear filtro"
    End Sub

    Shared Function CreateInstance(Of T)(filter As BaseFilter(Of T), properties As Dictionary(Of String, PropertyInfo)) As FilterViewModel
        Dim vm As New FilterViewModel(properties)
        vm.SelectedPropertyIndex = vm.PropertiesNames.IndexOf(filter.DisplayPropertyName)
        vm.IsCreatingFilter = False
        vm.DialogTitle = "Editar filtro"
        Return vm
    End Function

    Protected Overrides Sub SetValidators()

    End Sub

    Protected Overrides Sub SetCommandUpdates()

    End Sub

    ReadOnly Property PropertiesNames As List(Of String)
        Get
            Dim list = _Properties.Keys.ToList
            list.Sort()
            Return list
        End Get
    End Property

    ReadOnly Property EnumValues As List(Of String)
        Get
            If FilterType.IsEnum Then
                Return EnumDescriptionsToList(FilterType)
            End If
            Return Nothing
        End Get
    End Property

    ReadOnly Property MathFunctionsNames As List(Of String)
        Get
            Return EnumDescriptionsToList(GetType(NumericFunctions))
        End Get
    End Property

    Property SelectedPropertyIndex As Integer
        Get
            Return _SelectedPropertyIndex
        End Get
        Set
            _SelectedPropertyIndex = Value
            OnPropertySelected()
            NotifyPropertyChanged(NameOf(SelectedPropertyIndex))
        End Set
    End Property

    Private Sub OnPropertySelected()
        PropertyDisplayName = PropertiesNames(SelectedPropertyIndex)
        PropertyName = _Properties(PropertyDisplayName).Name
        FilterType = _Properties(PropertyDisplayName).PropertyType
        If IsNumericType(FilterType) Then
            ShowNumericPanel = True
            ShowStringPanel = False
            ShowDatePanel = False
            ShowBooleanPanel = False
            ShowEnumPanel = False
            IsIntegerNumericType = Base.IsIntegerNumericType(FilterType)
        Else
            If FilterType.IsEnum Then
                ShowNumericPanel = False
                ShowStringPanel = False
                ShowDatePanel = False
                ShowBooleanPanel = False
                ShowEnumPanel = True
                NotifyPropertyChanged(NameOf(EnumValues))
            ElseIf FilterType Is GetType(String) Then
                ShowNumericPanel = False
                ShowStringPanel = True
                ShowDatePanel = False
                ShowBooleanPanel = False
                ShowEnumPanel = False
            ElseIf FilterType Is GetType(Date) Then
                ShowNumericPanel = False
                ShowStringPanel = False
                ShowDatePanel = True
                ShowEnumPanel = False
                ShowBooleanPanel = False
            ElseIf FilterType Is GetType(Boolean) Then
                ShowNumericPanel = False
                ShowStringPanel = False
                ShowDatePanel = False
                ShowEnumPanel = False
                ShowBooleanPanel = True
            Else
                View.ShowErrorMessageBox("No se puede filtrar por esta propiedad.")
            End If
        End If

        NotifyPropertyChanged(NameOf(ShowDatePanel))
        NotifyPropertyChanged(NameOf(ShowStringPanel))
        NotifyPropertyChanged(NameOf(ShowNumericPanel))
        NotifyPropertyChanged(NameOf(ShowBooleanPanel))
        NotifyPropertyChanged(NameOf(ShowEnumPanel))
    End Sub

End Class

File: Wells/src/ViewModels/FilterViewModel.cs
using Wells.BaseView.Validators;
using Wells.BaseView.ViewModel;
using Wells.Base;
using Wells.View.Filters;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using System.Windows.Input;
using Wells.BaseView;

namespace Wells.View.ViewModels
{
    public class FilterViewModel : BaseViewModel
    {
        private int _SelectedPropertyIndex;
        private readonly Dictionary<string, PropertyInfo> _Properties;

        public bool ShowStringPanel { get; set; }
        public bool ShowNumericPanel { get; set; }
        public bool ShowBooleanPanel { get; set; }
        public bool ShowDatePanel { get; set; }
        public bool ShowEnumPanel { get; set; }
        public double NumericValue { get; set; }
        public string StringValue { get; set; }
        public int SelectedEnumValue { get; set; }
        public DateTime StartDate { get; set; }
        public DateTime EndDate { get; set; }
        public bool BooleanValue { get; set; }
        public Type FilterType { get; set; }
        public int SelectedMathFunction { get; set; }
        public string PropertyName { get; set; }
        public string PropertyDisplayName { get; set; }
        public bool IsIntegerNumericType { get; set; }
        public bool IsCreatingFilter { get; set; }
        public string DialogTitle { get; set; }
        public List<string> PropertiesNames
        {
            get
            {
                var list = _Properties.Keys.ToList();
                list.Sort();
                return list;
            }
        }

        public List<string> EnumValues
        {
            get
            {
                if (FilterType.IsEnum)
                {
                    return Common.EnumDescriptionsToList(FilterType);
                }
                return null;
            }
        }

        public List<string> MathFunctionsNames
        {
            get
            {
                return Common.EnumDescriptionsToList(typeof(NumericFunctions));
            }
        }

        public int SelectedPropertyIndex
        {
            get => _SelectedPropertyIndex;
            set
            {
                _SelectedPropertyIndex = value;
                OnPropertySelected();
                NotifyPropertyChanged(nameof(SelectedPropertyIndex));
            }
        }

        public FilterViewModel(Dictionary<string, PropertyInfo> properties) : base(null)
        {
            _Properties = properties;
            EndDate = DateTime.Today;
            BooleanValue = true;
            SelectedEnumValue = 0;
            IsIntegerNumericType = false;
            IsCreatingFilter = true;
            if (properties.Any())
            {
                SelectedPropertyIndex = 0;
            }
            DialogTitle = "Crear filtro";
        }

        public static FilterViewModel CreateInstance<T>(BaseFilter<T> filter, Dictionary<string, PropertyInfo> properties)
        {
            var vm = new FilterViewModel(properties);
            vm.SelectedPropertyIndex = vm.PropertiesNames.IndexOf(filter.DisplayPropertyName);
            vm.IsCreatingFilter = false;
            vm.DialogTitle = "Editar filtro";
            return vm;
        }

        private void OnPropertySelected()
        {
            PropertyDisplayName = PropertiesNames[SelectedPropertyIndex];
            PropertyName = _Properties[PropertyDisplayName].Name;
            FilterType = _Properties[PropertyDisplayName].PropertyType;
            if (Common.IsNumericType(FilterType))
            {
                ShowNumericPanel = true;
                ShowStringPanel = false;
                ShowDatePanel = false;
                ShowBooleanPanel = false;
                ShowEnumPanel = false;
                IsIntegerNumericType = Common.IsIntegerNumericType(FilterType);
            }
            else
            {
                if (FilterType.IsEnum)
                {
                    ShowNumericPanel = false;
                    ShowStringPanel = false;
                    ShowDatePanel = false;
                    ShowBooleanPanel = false;
                    ShowEnumPanel = true;
                    NotifyPropertyChanged(nameof(EnumValues));
                }
                else if (FilterType == typeof(string))
                {
                    ShowNumericPanel = false;
                    ShowStringPanel = true;
                    ShowDatePanel = false;
                    ShowBooleanPanel = false;
                    ShowEnumPanel = false;
                }
                else if (FilterType == typeof(DateTime))
                {
                    ShowNumericPanel = false;
                    ShowStringPanel = false;
                    ShowDatePanel = true;
                    ShowEnumPanel = false;
                    ShowBooleanPanel = false;
                }
                else if (FilterType == typeof(bool))
                {
                    ShowNumericPanel = false;
                    ShowStringPanel = false;
                    ShowDatePanel = false;
                    ShowEnumPanel = false;
                    ShowBooleanPanel = true;
                }
                else
                {
                    View.ShowErrorMessageBox("No se puede filtrar por esta propiedad.");
                }
            }

            NotifyPropertyChanged(nameof(ShowDatePanel));
            NotifyPropertyChanged(nameof(ShowStringPanel));
            NotifyPropertyChanged(nameof(ShowNumericPanel));
            NotifyPropertyChanged(nameof(ShowBooleanPanel));
            NotifyPropertyChanged(nameof(ShowEnumPanel));
        }


        public ICommand SaveFilterCommand
        {
            get
            {
                return new RelayCommand((param) =>
                {
                    CloseModalViewCommand.Execute(true);
                }, (obj) => IsValidFilter(), OnError);
            }
        }

        bool IsValidFilter()
        {
            if (ShowStringPanel)
            {
                return !string.IsNullOrEmpty(StringValue);
            }
            return true;
        }

        protected override void SetCommandUpdates() {
            Add(nameof(StringValue), SaveFilterCommand);
            Add(nameof(ShowStringPanel), SaveFilterCommand);
        }

        protected override void SetValidators() {
           //Nothing to add yet.
        }
    }
}
